### DISCUSSION TITLE
@next/third-parties Google Analytics compliance

### DISCUSSION THREAD

**Discussion by wuguishifu** [vercel/next.js#67440] [Help] [answered]
### Summary

Is the `<GoogleAnalytics />` component from `@next/third-parties/google` GDPR and CCPA compliant?

### Additional information

It seems as though the general instructions are to add the `<GoogleAnalytics />` component in your `<html />` in your root layout like this:

```tsx
export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang='en'>
      <body>
        {children}
      </body>
      <GoogleAnalytics gaId='G-********' />
    </html>
  );
};
```

Is there no way to conditionally disable tracking or wait for tracking to occur after the user has accepted consent? Otherwise, I don't see how this library can possibly be GDPR and CCPA compliant. Putting it in the root layout means we can't evaluate if the user has consented to trackers (would be stored as a cookie).

Has anyone found a solution to this? I don't want to have to use the `<Script />` method because frankly it's a PITA but it beats getting into a GDPR lawsuit!

**Comment by versecafe** [comment, marked as answer]
@wuguishifu You can wrap your conditional logic around accepting the privacy policy and if GA is allowed in a component on the client which then off that renders GoogleAnalytics or not I recommend using local storage to track this so it doesn't pop up every time they open the page, there is some additional information on ways to implement things here https://github.com/vercel/next.js/discussions/64985

**Reply by versecafe** [reply]
```tsx
if(gaAllowed) {
  return <GoogleAnalytics />
} else {
  return <></>
}
```
use a useState to control this with a default of false

**Reply by konhi** [reply]
This is not optimal solution since it doesn't use no-consent mode, Analytics shouldn't load only after consent is given. Using [c15t](https://c15t.com/docs/integrations/google-tag) is probably the best bet right now. Also worth looking at https://cookiebench.com/

**Comment by wuguishifu** [comment]
Oh, okay thanks. I wasn't aware that I could use the analytics in a client component. Thanks!

**Comment by dpasirst** [comment]
I could be wrong as I'm not a GA expert but that answer does not seem completely correct. Google Analytics (and Google Tag Manager) should function (with lesser details) without the consent of cookies. However, the code above would completely disable all of GA's functionality.

There are some details on [Consent Mode](https://developers.google.com/tag-platform/security/guides/consent?consentmode=advanced) as supported by Google - but to @wuguishifu original question, it is not clear how to access, set, use the consent mode capabilities with @next/third-parties.

As a side note, unfortunately not using @next/third-parties, someone [wrote this blog](https://dev.to/idboussadel/implementing-google-analytics-in-consent-mode-with-a-cookie-banner-for-nextjs-with-ts-1ga6) which does seem to work as expected (lesser GA details without cookies, full details with cookies).

**Reply by julian100004** [reply]
Came across this as i have the same problem right now. 
Is there a solution for this now except either disabling the Tag Manager completely or not using `@next/third-parties` and going the "old" route?

**Reply by dpasirst** [reply]
I ended up going the "old" route per my comment as a clean approach with `@next/third-parties` was not obvious to me.

**Reply by julian100004** [reply]
Alright, thank you for the reply. I'll go for the same route then 👍👍

**Reply by OliverFluid** [reply]
I funnily enough came to the same conclusion, usually we would do the oldschool way, but thought I would reach for the third-parties library this time. But seems theres no obvious way to handle consent correctly as you're correct the analytics/gtm should mount but in consent mode.

**Reply by ggralak** [reply]
Question is if the third-parties Google Analytics component sets **analytics_storage** to **denied**. If so then tracking is disabled by default and once a visitor to the website gives you the consent you can set **analytics_storage** to **granted**.

`window.gtag("consent", "update", {
                analytics_storage: "granted",
            });`

**Reply by dpasirst** [reply]
> Question is if the third-parties Google Analytics component sets **analytics_storage** to **denied**. If so then tracking is disabled by default and once a visitor to the website gives you the consent you can set **analytics_storage** to **granted**.
> 
> `window.gtag("consent", "update", { analytics_storage: "granted", });`

I'm not sure what you mean by `"tracking"` - my understanding (which could be wrong) is that settings values like:
```
ad_storage, ad_user_data, ad_personalization, analytics_storage
```
to `denied` would prevent cookies, while setting to `granted` would allow cookies from GA (is this what you mean about tracking?). Even with cookies denied, it is possible to collect limited analytics in GA for your site regarding the visit from the user.

**Comment by kaifirschau** [comment]
Is there any news on this? I would like to use the package from the next/third-parties provider, but I need to ensure full GDPR compliance.

**Reply by konhi** [reply]
What "news" are you expecting?

**Reply by konhi** [reply]
IMO, use c15t
- most performant - https://cookiebench.com
- supports Google Analytics with no-consent mode out-of-the-box - https://c15t.com/docs/integrations/google-tag

**Reply by kaifirschau** [reply]
@konhi that was exactly the kind of "good-solution-for-this-problem"-news I was expecting! :D thank you very much!
